bpf/lib/encrypt.h

```diff
@@ -205,6 +206,179 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	if (ip_proto == IPPROTO_ESP)
+		return 0;
+	if (src_sec_id == HOST_ID)
+		return 0;
+	if (dst_sec_id == HOST_ID)
+		return 0;
+	if (!identity_is_cluster(dst_sec_id))
+		return 0;
+	if (!identity_is_cluster(src_sec_id))
+		return 0;
+	if (identity_is_remote_node(dst_sec_id))
+		return 0;
+	if (identity_is_remote_node(src_sec_id))
+		return 0;
+	return 1;
+}
+
+static __always_inline int
+ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
+{
+	struct remote_endpoint_info __maybe_unused *dst = NULL;
+	struct remote_endpoint_info __maybe_unused *src = NULL;
+	void *data __maybe_unused, *data_end __maybe_unused;
+	struct iphdr __maybe_unused *ip4;
+	struct iphdr __maybe_unused *ip4_inner;
+	struct ipv6hdr __maybe_unused *ip6;
+	__u32 l4_off __maybe_unused = 0;
+	__u32 magic __maybe_unused = 0;
+	int ip_proto = 0;
+	int ret = 0;
+	union macaddr dst_mac = CILIUM_NET_MAC;
+
+	if (!eth_is_supported_ethertype(proto))
+		return DROP_UNSUPPORTED_L2;
+
+	switch (proto) {
+# ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+
+		ip_proto = ip4->protocol;
+
+		dst = lookup_ip4_remote_endpoint(ip4->daddr, 0);
+		src = lookup_ip4_remote_endpoint(ip4->saddr, 0);
+
+#  if defined(TUNNEL_MODE)
+		/* tunnel mode needs a bit of special handling when
+		 * encapsulated packets get here the destination address is
+		 * already a cluster node IP.
+		 *
+		 * the security ID is appended to the mark in the overlay prog
+		 * and we can extract this with 'get_identity'.
+		 * additionally, this is a VXLAN packet so ip4->daddr is the ip
+		 * of the destination host already and can be passed into
+		 * set_ipsec_encrypt to obtain the correct node ID and spi.
+		 */
+		if (ctx_is_overlay(ctx)) {
+			/* if bpf_overlay is v1.17 we immediately know if the
+			 * overlay prog saw this packet as encrypted, short
+			 * circuit.
+			 */
+			if (ctx_is_overlay_encrypted(ctx))
+				return CTX_ACT_OK;
+
+			/* in v1.17 backport code we may see in-flight packets
+			 * processed by v1.16 to_overlay program which does
+			 * not mark the packets for the subsequent
+			 * "ctx_is_overlay_encrypted" check.
+			 *
+			 * in this scenario we must parse headers to determine
+			 * if the tunnel traffic is encapsulated ESP and should
+			 * be encrypted per v1.17 IPsec rules.
+			 *
+			 * we can reuse the vxlan related function for geneve
+			 * as well since IPsec does not currently support DSR,
+			 * meaning both VXLan and Geneve headers are 8 bytes
+			 * large with no additional options for Geneve.
+			 */
+			l4_off = ETH_HLEN + ipv4_hdrlen(ip4);
+			if (!vxlan_get_inner_ipv4(data, data_end, l4_off,
+						  &ip4_inner))
+					return CTX_ACT_OK;
+
+			/* if this is already encrypted, just pass it to stack */
+			if (ip4_inner->protocol == IPPROTO_ESP)
+				return CTX_ACT_OK;
+
+			/* determine if we need should encrypt the non-ESP
+			 * tunnel traffic per v1.17 rules
```

Comment:
@ldelossa we also need to handle an embedded IPv6 packet here, at the very least.

Reply:
Ahhhh good catch. 

Okay, so this little bit actually needs its own protocol switch. 
I will need to write a bit more code here to look at the inner eth and determine the next IP protocol

Reply:
https://github.com/cilium/cilium/pull/38843